src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1095,6 +1127,30 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
 
                     // Eat exception and try again.
                 }
+                finally
+                {
+                    CancelWithGlobalTimeout(this, http11ConnectionWaiter);
+                    CancelWithGlobalTimeout(this, http2ConnectionWaiter);
```

Comment:
In what situation would both of these be in need of cancellation?  If you nulled them out upon successful completion of waiting for the request, could this instead be:
```C#
CancelWithGlobalTimeout(this, http11ConnectionWaiter ?? http2ConnectionWaiter);
```
?

Also, as a nit, can we name this CancelIfNecessary or something like that?  It's not always going to use a timeout, and there may not even be anything to cancel.

Reply:
> In what situation would both of these be in need of cancellation

When downgrade happens, it's possible that `http11ConnectionWaiter != null && http2ConnectionWaiter != null`. In that case `http2ConnectionWaiter.ConnectionCancellationTokenSource` will be `null`, so we never need to cancel both. It didn't seem to worth explicitly nulling out `http2ConnectionWaiter` in case of downgrade. IMO we should just rename the method as you suggest and add a comment.